sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousRateStreamSource.scala

```diff
@@ -0,0 +1,155 @@
+
+  override def setOffset(offset: java.util.Optional[Offset]): Unit = {
+    this.offset = offset
+  }
+
+  override def getStartOffset(): Offset = offset.get()
+
+  override def createReadTasks(): java.util.List[ReadTask[Row]] = {
+    val partitionStartMap = Option(offset.orElse(null)).map {
+      case o: ContinuousRateStreamOffset => o.partitionToStartValue
+      case s: SerializedOffset => Serialization.read[Map[Int, Long]](s.json)
+      case _ => throw new IllegalArgumentException("invalid offset type for ContinuousRateSource")
+    }
+    if (partitionStartMap.exists(_.keySet.size > numPartitions)) {
+      throw new IllegalArgumentException("Start offset contained too many partitions.")
+    }
+    val perPartitionRate = rowsPerSecond.toDouble / numPartitions.toDouble
+
+    Range(0, numPartitions).map { n =>
+      // If the offset doesn't have a value for this partition, start from the beginning.
+      val start = partitionStartMap.flatMap(_.get(n)).getOrElse(0L + n)
+      // Have each partition advance by numPartitions each row, with starting points staggered
+      // by their partition index.
+      RateStreamReadTask(start, n, numPartitions, perPartitionRate)
+        .asInstanceOf[ReadTask[Row]]
+    }.asJava
+  }
+
+  override def commit(end: Offset): Unit = {}
+  override def stop(): Unit = {}
+
+}
+
+case class RateStreamReadTask(
+    startValue: Long, partitionIndex: Int, increment: Long, rowsPerSecond: Double)
+  extends ReadTask[Row] {
+  override def createDataReader(): DataReader[Row] =
+    new RateStreamDataReader(startValue, partitionIndex, increment, rowsPerSecond.toLong)
+}
+
+class RateStreamDataReader(
+    startValue: Long, partitionIndex: Int, increment: Long, rowsPerSecond: Long)
+  extends ContinuousDataReader[Row] {
+  private var nextReadTime = 0L
+  private var numReadRows = 0L
+
+  private var currentValue = startValue
+  private var currentRow: Row = null
+
+  override def next(): Boolean = {
+    // Set the timestamp for the first time.
+    if (currentRow == null) nextReadTime = System.currentTimeMillis() + 1000
+
+    if (numReadRows == rowsPerSecond) {
+      // Sleep until we reach the next second.
+
+      try {
+        while (System.currentTimeMillis < nextReadTime) {
+          Thread.sleep(nextReadTime - System.currentTimeMillis)
+        }
+      } catch {
+        case _: InterruptedException =>
+          // Someone's trying to end the task; just let them.
+          return false
+      }
+      numReadRows = 0
+      nextReadTime += 1000
+    }
+
+    currentRow = Row(
+      DateTimeUtils.toJavaTimestamp(DateTimeUtils.fromMillis(System.currentTimeMillis)),
+      currentValue)
+    currentValue += increment
+    numReadRows += 1
+
+    true
+  }
+
+  override def get: Row = currentRow
+
+  override def close(): Unit = {}
+
+  override def getOffset(): PartitionOffset =
+    ContinuousRateStreamPartitionOffset(partitionIndex, currentValue - increment)
```

Comment:
`currentValue - increment` will be wrong if the first `next()` call is interrupted.

Reply:
(Your comment made me relalize there was a fencepost error here, which I'll swap some things around to fix.)

getOffset(), like get(), is supposed to be paired with next() in a pseudo-iterator and non-thread-safe manner. It'd be incorrect for a caller to look for getOffset() to be correct after a failed next() call.